_posts/2015-03-stitch-fix-heart-unix.md

```diff
@@ -0,0 +1,639 @@
+
+#### Problem
+Some directory contains a large _number_ of files (they don't take up a lot of disk space) and I want to find which one
+
+#### Solution
+`find` command within a bash loop similar to the du command.
+
+##### Input
+`cd` to some directory
+
+##### Command
+```
+find . -type d | while read -r dir;
+  do printf "%d\t%s\n" `find "$dir" | wc -l` ""$dir"";
+done | sort -n
+```
+
+##### Output
+```
+1	./IPython-notebook-extensions/.git/objects/info
+1	./IPython-notebook-extensions/.git/refs/tags
+...
+1134	./sf/flinch
+4273	./sf
+5709	.
+```
+
+ - The first `find` gets all the directories below the current one
+ - The while loop goes over each directory and finds all the files below it.  
+   - Not efficient, but I haven't yet run into situations where it takes too long.  
+ - The final sort command puts output in a useful order.  
+   - The number of files is _cumulative_
+   - In the above example, there are 4273 files in _all_ directories below ./sf
+
+<a name="delete-directory-with-large-number-of-files"></a>
+### Delete Directory With Large Number Of Files
+&#8212; [Greg Novak](https://www.linkedin.com/in/gsnovak/) *(OS X / bash)*
+
+#### Problem
+I have a directory containing a large number of files and I want to delete it, but 'rm *' gives "Argument list too long" and refuses.
+
+#### Solution
+`xargs` command
+
+##### Input
+`cd` to some directory
+
+##### Command
+```
+# Replace ls with rm to delete
+find . | xargs -n 100 ls
+```
+
+- `xargs` will execute the given command on batches of 100 files at a time.
+- Note the replacement of the typical `rm` with `ls`
+
+<a name="csvs"></a>
+## CSVs
+
+<a name="filter-csv-by-column-values"></a>
+### Filter CSV File By Column Values
+&#8212; [Simeon Willbanks](https://twitter.com/simeonwillbanks/) *(OS X / zsh)*
+
+#### Problem
+We must filter a CSV file by specific column values.
+
+#### Solution
+`awk` the csv!
+
+##### Input
+`./hours.csv`
+
+```csv
+User ID,Hours Styling
+1,15.90
+2,17.43
+3,15.01
+4,18.20
+5,15.55
+6,16.33
+```
+
+##### Command
+```
+awk -F, '$2 ~ 15' ./hours.csv
+```
+
+##### Output
+```
+1,15.90
+3,15.01
+5,15.55
+```
+
+ - `awk` `-F` sets a "field separator"; for CSV files, this is a ','
+ - `'$2 ~ 15'` is the `awk` program
+   - `$2` is the second field which is "Hours Styling"
+   - `~` is a regular expression operator, so we filter any lines with hours that match `15` 
+
+##### Command
+```
+awk -F, '$2 == 15' ./hours.csv
+```
+
+##### Output
+```
+3,15.01
+```
+
+ - `==` is an equality operator, so we filter any lines with hours that match `15.01` 
+
+<a name="csv-file-columns-names-and-indices"></a>
+### CSV File Column Names And Indices
+&#8212; [Nick Kridler](https://github.com/nmkridler/) *(OS X, Linux, Unix / bash)*
+
+#### Problem
+We don't know a CSV files column names and indices.
+
+#### Solution
+Use `head` to get the header and pipe it into `awk`
+
+##### Input
+`./file.csv`
+
+```csv
+a,b,c,d
```

Comment:
This could stand to be a more real-world example.  The a's and 1's run together making it hard to understand what I'm looking at, e.g. is the "1" in the output the "1" from this CSV?